### ISSUE TITLE
Plans for Node Group Management and Service Scope Management

### ISSUE THREAD

**Issue by vincentgoat** [kubeedge/kubeedge#3582]
Hi guys, we have plans for achieving the feature of Pod Scheduling among node groups in phase 1, and we also make plans for achieving the feature of Service Scope within a node group in phase 2.

If you have features to achieve or to get involved in the development, please see the project form below, and you can add other todos here as you want:
**Phases 1: Node Group Management**
- [ ] Define a NodeGroup CR to indicate which nodes belong to the node group.
- [ ] Refer to Karmada policy, define a PropagationPolicy CR that specifies which node group pods should be scheduled to and how many pods should run in this node group.
- [ ] Unit/E2E Test

**Phases 2: Service Scope Management**
TODO

**Comment by vincentgoat** [comment]
2022.1.24 Meeting Minutes:
1. Achieve the override in cloud core, not only override images path but also other attributes, the data comes from an independent policy.
2. Use the kubectl plug-in to new operations for node group.
3. The controller runtime version depends on the Kubeedge version update.
4. Determine the API.
5. Create a Google Doc for more details about the proposal.

**Comment by vincentgoat** [comment]
proposal: #3574

**Comment by Congrool** [comment]
2022.1.26 Meeting:
1. Only focus on new applications which have not been deployed.
2. Find some way to solve the contention when applying propogation policy, such as make pods with specified label/annotation in pending state until the extender finds the relative policy.
3. Modify the PropagationPolicy API to support  scheduling explicit number of pods to specified nodegroups.

**Comment by benjaminhuo** [comment]
Looks like NodeGroup and ServiceScope is trying to achieve goals similar to [OpenYurt YurtAppManager](https://openyurt.io/docs/core-concepts/yurt-app-manager) and [SuperEdge ServiceGroup](https://superedge.io/docs/components/servicegroup/)  but in different ways?

Both [OpenYurt YurtAppManager](https://openyurt.io/docs/core-concepts/yurt-app-manager) and [SuperEdge ServiceGroup](https://superedge.io/docs/components/servicegroup/)  distribute workload by creating sub-deployment while KubeEdge is try to distribute pods in a single deployment to different NodeGroup?

It might work well for deployment, but is this ok for statefulset to spread pods of a single statefulset to different edge locations?  For example: 
- 2 pods for a MySQL statefulset, one is primary and another is replica
- 3 pods for an Elasticsearch statefulset, each pod holds a portion of the data

And where can I find the proposal for Service Scope?

Thanks
Ben

**Comment by Congrool** [comment]
@benjaminhuo Thanks for your feedback.
> It might work well for deployment, but is this ok for a statefulset to spread pods of a single statefulset to different edge locations?

It's the problem that we need take into consideration. This is not the final design and any other suggestions are welcome.

> And where can I find the proposal for Service Scope?

Do you mean Serivce Scope based on something like node group? And a pod can only reach endpoints in the same nodegroup. In my mind, it is still under designing.

**Comment by benjaminhuo** [comment]
> Do you mean Serivce Scope based on something like node group? And a pod can only reach endpoints in the same nodegroup. In my mind, it is still under designing.

Yep, I mean one multi-nodegroup deployment should only have one service, and the access to this service from a pod in the same nodegroup should only be able to find endpoints in the same nodegroup.

**Comment by Congrool** [comment]
Well, because KubeEdge doesn't have the concept of node group before this proposal, I think the proposal of service scope based on it haven't been post yet. You can watch [the edgemesh repo](https://github.com/kubeedge/edgemesh) which mantains the network solutions of KubeEdge and any new information will be post there.
@benjaminhuo

**Comment by fisherxu** [comment]
@benjaminhuo  I personally tend to wrap the workload in a CRD, then distribute the workload to the node group.

>  I mean one multi-nodegroup deployment should only have one service, and the access to this service from a pod in the same nodegroup should only be able to find endpoints in the same nodegroup.

You mean only create one service for all the node groups? That maybe leads to the pods(ip) from different group located in one endpointSlice. Hard to seprated for groups.

**Comment by benjaminhuo** [comment]
> You mean only create one service for all the node groups? That maybe leads to the pods(ip) from different group located in one endpointSlice. Hard to seprated for groups.

My previous conclusion is based on the current proposal because there is only one deployment for multiple node groups in the current proposal.

Current design might work for deployment, but not for statefulset and might have issues for service.

**Comment by fisherxu** [comment]
As we discussed lately, for the workload like deployment, we decide to deploy sub-workload to every group through CRD, so every group will have a workload(deployment).

But for the service, if we create only one for all workload, now in k8s there're two features related to the topology(worked with kube-proxy):
1. Service topology keys: It can keep the traffic within a zone. But the feature is deprecated and removed now.
2. Service topology hints: The new version of service topology. But it can't ensure keeping the traffic within a zone, it's the best-effort strategy currently, maybe not meet our needs.

Another idea is to create one service for every nodegroup, which can solve the problem of isolation, but there will be to many svcs. 

And we also need the gateway/ingress for every node group.

How do you think? @benjaminhuo @Congrool @vincentgoat  @zc2638

**Comment by Congrool** [comment]
@fisherxu 
I'm not be familiar with edgemesh, so it's just my thought:
Beacause edgemesh takes over the work of kube-proxy as well as coreDNS, can we make the edgemesh aware of the nodegroup? So that edgemesh can remain endpoints in the same nodegroup and discard other endpoints that not. In this way, only iptables rules for endpoints in the same nodegroup will be set and the dns will only be resolved to ips of endpoints in the same nodegroup.

Is this implementable?

**Comment by fisherxu** [comment]
> Beacause edgemesh takes over the work of kube-proxy as well as coreDNS, can we make the edgemesh aware of the nodegroup?

Yes, EdgeMesh is one networking solution for us, and will do this :) 
And I think we still need to consider to compatible with other solutions as kube-proxy even service mesh like istio.

**Comment by benjaminhuo** [comment]
I agree that EdgeMesh should be aware of the nodegroup.

As regarding the kube-proxy approach, OpenYurt is using [EndpointSliceProxying](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/) together with topologyKeys to use one service for all sub deployment in all nodepool. https://github.com/openyurtio/openyurt/blob/master/docs/tutorial/service-topology.md

But it requires:
- kube-proxy needs to be configured to connect to Yurthub instead of the API Server.
- restart kube-proxy

@fisherxu @Congrool

**Comment by vincentgoat** [comment]
EdgeApplication proposal: [https://docs.google.com/document/d/19FHzmHB-a-OSUukmicEVga6neMCeHz_ozPFxaGXqJ1I/edit#heading=h.d5vhf8rdz3fj](url)

**Comment by gy95** [comment]
> EdgeApplication proposal: [https://docs.google.com/document/d/19FHzmHB-a-OSUukmicEVga6neMCeHz_ozPFxaGXqJ1I/edit#heading=h.d5vhf8rdz3fj](url)


no authority :)

**Comment by benjaminhuo** [comment]
Yeah, this doc is not open yet

**Comment by vincentgoat** [comment]
2022.3.2 meeting：
1、Readability optimization for unified packaging;
2、Consider the EdgeApplication concept in the cloud and edge side.

**Comment by vincentgoat** [comment]
> > EdgeApplication proposal: [https://docs.google.com/document/d/19FHzmHB-a-OSUukmicEVga6neMCeHz_ozPFxaGXqJ1I/edit#heading=h.d5vhf8rdz3fj](url)
> 
> no authority :)

Plz join the group kubeedge@googlegroups.com

**Comment by vincentgoat** [comment]
2022.3.7 meeting:
1. Simplify EdgeApplication status field, the status show sub-resource summary status and where sub-resource located at;
2. Summary status contains progressing/available.

**Comment by koulq** [comment]
hi，In this design, have you ever considered the certificate management of device access?For example, with my same application deployment packaged in node groups A and node group B, do I require node groups A and node group B to use the same set of certificates issued by CA?

**Comment by vincentgoat** [comment]
> hi，In this design, have you ever considered the certificate management of device access?For example, with my same application deployment packaged in node groups A and node group B, do I require node groups A and node group B to use the same set of certificates issued by CA?

Hi @koulq , thanks for the feedback, we currently manage workload and service through nodegroup, and management of device certificates can be considered in future evolutions, and also you are welcome to follow up on this process.
